tests for Python 3.7 started failing with 'TypeError: Argument 'obj' has incorrect type (expected list, got FrozenList)'

The scheduled tests for Python 3.7 are failing now, see https://github.com/alchemistry/alchemlyb/runs/2983317549?check_suite_focus=true

Lots of errors like to following:

```
 ______________________________ Test_Units.test_ti ______________________________

self = <alchemlyb.tests.test_ti_estimators.Test_Units object at 0x7f8ae3af9050>
dhdl =                           fep
time    fep-lambda           
0.0     0.0         13.390058
10.0    0.0          9.23137...338
39980.0 1.0         -1.184511
39990.0 1.0         -1.273484
40000.0 1.0         -0.416037

[20005 rows x 1 columns]

    def test_ti(self, dhdl):
>       ti = TI().fit(dhdl)

src/alchemlyb/tests/test_ti_estimators.py:177: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/alchemlyb/estimators/ti_.py:64: in fit
    dl = means.reset_index()[means.index.names[:]].diff().iloc[1:].values
/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/pandas/core/frame.py:3445: in __getitem__
    if com.is_bool_indexer(key):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

key = FrozenList(['fep-lambda'])

    def is_bool_indexer(key: Any) -> bool:
        """
        Check whether `key` is a valid boolean indexer.
    
        Parameters
        ----------
        key : Any
            Only list-likes may be considered boolean indexers.
            All other types are not considered a boolean indexer.
            For array-like input, boolean ndarrays or ExtensionArrays
            with ``_is_boolean`` set are considered boolean indexers.
    
        Returns
        -------
        bool
            Whether `key` is a valid boolean indexer.
    
        Raises
        ------
        ValueError
            When the array is an object-dtype ndarray or ExtensionArray
            and contains missing values.
    
        See Also
        --------
        check_array_indexer : Check that `key` is a valid array to index,
            and convert to an ndarray.
        """
        if isinstance(key, (ABCSeries, np.ndarray, ABCIndex)) or (
            is_array_like(key) and is_extension_array_dtype(key.dtype)
        ):
            if key.dtype == np.object_:
                key = np.asarray(key)
    
                if not lib.is_bool_array(key):
                    na_msg = "Cannot mask with non-boolean array containing NA / NaN values"
                    if lib.infer_dtype(key) == "boolean" and isna(key).any():
                        # Don't raise on e.g. ["A", "B", np.nan], see
                        #  test_loc_getitem_list_of_labels_categoricalindex_with_na
                        raise ValueError(na_msg)
                    return False
                return True
            elif is_bool_dtype(key.dtype):
                return True
        elif isinstance(key, list):
            # check if np.array(key).dtype would be bool
>           return len(key) > 0 and lib.is_bool_list(key)
E           TypeError: Argument 'obj' has incorrect type (expected list, got FrozenList)
```

I don't know what the root problem is but perhaps something changed in pandas and we're picking up the latest version?

--- a/CHANGES
+++ b/CHANGES
@@ -25,6 +25,7 @@ Enhancements
 Deprecations
 
 Fixes
+  - Fix the support for pandas >= 1.3 (issue #147, PR #148).
 
 Changes
   - The scipy constant for gas constant is used for parsers instead of the

--- a/setup.py
+++ b/setup.py
@@ -43,5 +43,5 @@ setup(name='alchemlyb',
       long_description=open('README.rst').read(),
       long_description_content_type='text/x-rst',
       tests_require = ['pytest', 'alchemtest'],
-      install_requires=['numpy', 'pandas>=1.2', 'pymbar>=3.0.5,<4', 'scipy', 'scikit-learn', 'matplotlib']
+      install_requires=['numpy', 'pandas>=1.2,!=1.3.0', 'pymbar>=3.0.5,<4', 'scipy', 'scikit-learn', 'matplotlib']
       )

--- a/src/alchemlyb/estimators/ti_.py
+++ b/src/alchemlyb/estimators/ti_.py
@@ -1,5 +1,3 @@
-import pickle
-
 import numpy as np
 import pandas as pd
 

--- a/src/alchemlyb/estimators/ti_.py
+++ b/src/alchemlyb/estimators/ti_.py
@@ -1,3 +1,5 @@
+import pickle
+
 import numpy as np
 import pandas as pd
 
@@ -61,7 +63,7 @@ class TI(BaseEstimator):
         l_types = dHdl.index.names[1:]
 
         # obtain vector of delta lambdas between each state
-        dl = means.reset_index()[means.index.names[:]].diff().iloc[1:].values
+        dl = means.reset_index()[list(means.index.names[:])].diff().iloc[1:].values
 
         # apply trapezoid rule to obtain DF between each adjacent state
         deltas = (dl * (means.iloc[:-1].values + means.iloc[1:].values)/2).sum(axis=1)
@@ -133,7 +135,7 @@ class TI(BaseEstimator):
         # get the lambda names
         l_types = self.dhdl.index.names
         # obtain bool of changed lambdas between each state
-        lambdas = self.dhdl.reset_index()[l_types]
+        lambdas = self.dhdl.reset_index()[list(l_types)]
         diff = lambdas.diff().to_numpy(dtype='bool')
         # diff will give the first row as NaN so need to fix that
         diff[0, :] = diff[1, :]

--- a/src/alchemlyb/estimators/ti_.py
+++ b/src/alchemlyb/estimators/ti_.py
@@ -61,6 +61,8 @@ class TI(BaseEstimator):
         l_types = dHdl.index.names[1:]
 
         # obtain vector of delta lambdas between each state
+        # Fix issue #148, where for pandas != 1.3.0
+        # dl = means.reset_index()[means.index.names[:]].diff().iloc[1:].values
         dl = means.reset_index()[list(means.index.names[:])].diff().iloc[1:].values
 
         # apply trapezoid rule to obtain DF between each adjacent state
@@ -133,6 +135,8 @@ class TI(BaseEstimator):
         # get the lambda names
         l_types = self.dhdl.index.names
         # obtain bool of changed lambdas between each state
+        # Fix issue #148, where for pandas != 1.3.0
+        # lambdas = self.dhdl.reset_index()[l_types]
         lambdas = self.dhdl.reset_index()[list(l_types)]
         diff = lambdas.diff().to_numpy(dtype='bool')
         # diff will give the first row as NaN so need to fix that

--- a/src/alchemlyb/estimators/ti_.py
+++ b/src/alchemlyb/estimators/ti_.py
@@ -61,9 +61,9 @@ class TI(BaseEstimator):
         l_types = dHdl.index.names[1:]
 
         # obtain vector of delta lambdas between each state
-        # Fix issue #148, where for pandas != 1.3.0
-        # dl = means.reset_index()[means.index.names[:]].diff().iloc[1:].values
-        dl = means.reset_index()[list(means.index.names[:])].diff().iloc[1:].values
+        # Fix issue #148, where for pandas == 1.3.0
+        # dl = means.reset_index()[list(means.index.names[:])].diff().iloc[1:].values
+        dl = means.reset_index()[means.index.names[:]].diff().iloc[1:].values
 
         # apply trapezoid rule to obtain DF between each adjacent state
         deltas = (dl * (means.iloc[:-1].values + means.iloc[1:].values)/2).sum(axis=1)
@@ -135,9 +135,9 @@ class TI(BaseEstimator):
         # get the lambda names
         l_types = self.dhdl.index.names
         # obtain bool of changed lambdas between each state
-        # Fix issue #148, where for pandas != 1.3.0
-        # lambdas = self.dhdl.reset_index()[l_types]
-        lambdas = self.dhdl.reset_index()[list(l_types)]
+        # Fix issue #148, where for pandas == 1.3.0
+        # lambdas = self.dhdl.reset_index()[list(l_types)]
+        lambdas = self.dhdl.reset_index()[l_types]
         diff = lambdas.diff().to_numpy(dtype='bool')
         # diff will give the first row as NaN so need to fix that
         diff[0, :] = diff[1, :]

